File: lld/problems/task-scheduler/python/task_scheduler.py
"""Task Scheduler LLD — Python reference implementation.

In-memory, priority-based job scheduler with delayed execution and retries,
running on a fixed worker-thread pool (single process — see the HLD track
for a distributed cron/scheduler design). Mirrors go/taskscheduler.go.
"""
from __future__ import annotations

import heapq
import itertools
import threading
import time
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Callable, Dict, Optional


class Status(Enum):
    PENDING = auto()
    RUNNING = auto()
    RETRYING = auto()
    SUCCEEDED = auto()
    FAILED = auto()


class RetryPolicy:
    """Computes the delay (in seconds) before the (attempt+1)-th retry."""

    def next_delay(self, attempt: int) -> float:
        raise NotImplementedError


@dataclass(frozen=True)
class ExponentialBackoff(RetryPolicy):
    base: float
    max: float

    def next_delay(self, attempt: int) -> float:
        d = self.base
        for _ in range(attempt):
            d *= 2
            if d > self.max:
                return self.max
        return d


_DEFAULT_RETRY_POLICY = ExponentialBackoff(base=0.01, max=0.2)

# Job.task is a zero-arg callable that raises on failure, returns normally
# on success — mirrors Go's `func() error`.
Task = Callable[[], None]


@dataclass
class Job:
    id: str
    priority: int
    run_at: float  # time.monotonic()-based deadline
    task: Task
    max_retries: int = 0
    retry: Optional[RetryPolicy] = None
    attempts: int = field(default=0, init=False)


@dataclass
class JobResult:
    job_id: str
    status: Status
    attempts: int = 0
    error: Optional[BaseException] = None


@dataclass(order=True)
class _QueueEntry:
    """Min-heap entry ordered by run_at (earliest due first), tie-broken by
    higher priority first. `seq` is a tiebreaker so heapq never has to
    compare Job objects (which aren't orderable)."""

    sort_key: tuple = field(compare=True)
    seq: int = field(compare=True)
    job: Job = field(compare=False)


class Scheduler:
    """Runs submitted jobs on a fixed worker-thread pool, respecting
    priority, delayed run_at times, and per-job retry policies."""

    def __init__(self, workers: int):
        self._workers = workers
        self._lock = threading.Lock()
        self._heap: list = []
        self._seq = itertools.count()
        self._results: Dict[str, JobResult] = {}
        self._threads: list = []
        self._stopped = True

    def submit(self, job: Job) -> None:
        """Enqueues a job. Safe to call before or after start(), and
        concurrently from multiple threads."""
        if job.retry is None:
            job.retry = _DEFAULT_RETRY_POLICY
        with self._lock:
            # Negate priority so the heap (min-first) surfaces the highest
            # priority first among equal run_at times.
            heapq.heappush(self._heap, _QueueEntry((job.run_at, -job.priority), next(self._seq), job))
            self._results[job.id] = JobResult(job_id=job.id, status=Status.PENDING)

    def start(self) -> None:
        self._stopped = False
        self._threads = []
        for i in range(self._workers):
            t = threading.Thread(target=self._worker_loop, name=f"scheduler-worker-{i}", daemon=True)
            self._threads.append(t)
            t.start()

    def stop(self) -> None:
        self._stopped = True
        for t in self._threads:
            t.join()

    def _worker_loop(self) -> None:
        while not self._stopped:
            job = self._pop_due()
            if job is not None:
                self._execute(job)
            else:
                time.sleep(0.002)

    def _pop_due(self) -> Optional[Job]:
        """Removes and returns the highest-priority job whose run_at has
        already elapsed, or None if none is ready."""
        with self._lock:
            if not self._heap:
                return None
            if self._heap[0].job.run_at > time.monotonic():
                return None
            return heapq.heappop(self._heap).job

    def _execute(self, job: Job) -> None:
        with self._lock:
            self._results[job.id].status = Status.RUNNING

        err: Optional[BaseException] = None
        try:
            job.task()
        except BaseException as e:  # noqa: BLE001 - mirror Go's generic error
            err = e
        job.attempts += 1

        with self._lock:
            res = self._results[job.id]
            res.attempts = job.attempts
            if err is None:
                res.status = Status.SUCCEEDED
                res.error = None
                return
            res.error = err
            if job.attempts <= job.max_retries:
                res.status = Status.RETRYING
                delay = job.retry.next_delay(job.attempts - 1)
                job.run_at = time.monotonic() + delay
                heapq.heappush(self._heap, _QueueEntry((job.run_at, -job.priority), next(self._seq), job))
                return
            res.status = Status.FAILED

    def get_result(self, job_id: str) -> Optional[JobResult]:
        with self._lock:
            res = self._results.get(job_id)
            if res is None:
                return None
            return JobResult(job_id=res.job_id, status=res.status, attempts=res.attempts, error=res.error)

    def wait_for_result(self, job_id: str, timeout: float) -> Optional[JobResult]:
        """Polls until job_id reaches a terminal state (SUCCEEDED or FAILED)
        or the timeout (in seconds) elapses."""
        deadline = time.monotonic() + timeout
        while True:
            res = self.get_result(job_id)
            if res is not None and res.status in (Status.SUCCEEDED, Status.FAILED):
                return res
            if time.monotonic() > deadline:
                return None
            time.sleep(0.002)


if __name__ == "__main__":
    scheduler = Scheduler(workers=2)
    scheduler.start()

    ran = {"count": 0}

    def demo_task() -> None:
        ran["count"] += 1
        print("demo job executed")

    scheduler.submit(Job(id="demo", priority=1, run_at=time.monotonic(), task=demo_task))

    result = scheduler.wait_for_result("demo", timeout=1.0)
    print(f"Result: {result}")

    scheduler.stop()

File: lld/problems/task-scheduler/go/taskscheduler.go
// Package taskscheduler implements an in-memory, priority-based job scheduler
// with delayed execution and retries (single process — see the HLD track for
// a distributed cron/scheduler design).
package taskscheduler

import (
	"container/heap"
	"sync"
	"time"
)

type Task func() error

type Status int

const (
	Pending Status = iota
	Running
	Retrying
	Succeeded
	Failed
)

// RetryPolicy computes the delay before the (attempt+1)-th retry.
type RetryPolicy interface {
	NextDelay(attempt int) time.Duration
}

type ExponentialBackoff struct {
	Base time.Duration
	Max  time.Duration
}

func (b ExponentialBackoff) NextDelay(attempt int) time.Duration {
	d := b.Base
	for i := 0; i < attempt; i++ {
		d *= 2
		if d > b.Max {
			return b.Max
		}
	}
	return d
}

type Job struct {
	ID         string
	Priority   int
	RunAt      time.Time
	Task       Task
	MaxRetries int
	Retry      RetryPolicy

	attempts int
}

type JobResult struct {
	JobID    string
	Status   Status
	Attempts int
	Err      error
}

// jobQueue is a min-heap ordered by RunAt (earliest due first), tie-broken by
// higher Priority first.
type jobQueue []*Job

func (q jobQueue) Len() int { return len(q) }
func (q jobQueue) Less(i, j int) bool {
	if q[i].RunAt.Equal(q[j].RunAt) {
		return q[i].Priority > q[j].Priority
	}
	return q[i].RunAt.Before(q[j].RunAt)
}
func (q jobQueue) Swap(i, j int) { q[i], q[j] = q[j], q[i] }
func (q *jobQueue) Push(x any)   { *q = append(*q, x.(*Job)) }
func (q *jobQueue) Pop() any {
	old := *q
	n := len(old)
	item := old[n-1]
	*q = old[:n-1]
	return item
}

var defaultRetryPolicy = ExponentialBackoff{Base: 10 * time.Millisecond, Max: 200 * time.Millisecond}

// Scheduler runs submitted jobs on a fixed worker pool, respecting priority,
// delayed RunAt times, and per-job retry policies.
type Scheduler struct {
	mu      sync.Mutex
	queue   jobQueue
	results map[string]*JobResult
	workers int
	stopCh  chan struct{}
	wg      sync.WaitGroup
}

func NewScheduler(workers int) *Scheduler {
	return &Scheduler{results: make(map[string]*JobResult), workers: workers}
}

// Submit enqueues a job. Safe to call before or after Start, and concurrently
// from multiple goroutines.
func (s *Scheduler) Submit(j *Job) {
	if j.Retry == nil {
		j.Retry = defaultRetryPolicy
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	heap.Push(&s.queue, j)
	s.results[j.ID] = &JobResult{JobID: j.ID, Status: Pending}
}

func (s *Scheduler) Start() {
	s.stopCh = make(chan struct{})
	for i := 0; i < s.workers; i++ {
		s.wg.Add(1)
		go s.workerLoop()
	}
}

func (s *Scheduler) Stop() {
	close(s.stopCh)
	s.wg.Wait()
}

func (s *Scheduler) workerLoop() {
	defer s.wg.Done()
	ticker := time.NewTicker(2 * time.Millisecond)
	defer ticker.Stop()
	for {
		select {
		case <-s.stopCh:
			return
		case <-ticker.C:
			if job := s.popDue(); job != nil {
				s.execute(job)
			}
		}
	}
}

// popDue removes and returns the highest-priority job whose RunAt has
// already elapsed, or nil if none is ready. Locking here (rather than around
// the whole execute) keeps job execution off the critical section so workers
// don't serialize on long-running tasks.
func (s *Scheduler) popDue() *Job {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.queue.Len() == 0 {
		return nil
	}
	if s.queue[0].RunAt.After(time.Now()) {
		return nil
	}
	return heap.Pop(&s.queue).(*Job)
}

func (s *Scheduler) execute(j *Job) {
	s.mu.Lock()
	s.results[j.ID].Status = Running
	s.mu.Unlock()

	err := j.Task()
	j.attempts++

	s.mu.Lock()
	defer s.mu.Unlock()
	res := s.results[j.ID]
	res.Attempts = j.attempts
	if err == nil {
		res.Status = Succeeded
		res.Err = nil
		return
	}
	res.Err = err
	if j.attempts <= j.MaxRetries {
		res.Status = Retrying
		delay := j.Retry.NextDelay(j.attempts - 1)
		j.RunAt = time.Now().Add(delay)
		heap.Push(&s.queue, j)
		return
	}
	res.Status = Failed
}

func (s *Scheduler) GetResult(id string) (JobResult, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	r, ok := s.results[id]
	if !ok {
		return JobResult{}, false
	}
	return *r, true
}

// WaitForResult polls until job id reaches a terminal state (Succeeded or
// Failed) or the timeout elapses.
func (s *Scheduler) WaitForResult(id string, timeout time.Duration) (JobResult, bool) {
	deadline := time.Now().Add(timeout)
	for {
		if r, ok := s.GetResult(id); ok && (r.Status == Succeeded || r.Status == Failed) {
			return r, true
		}
		if time.Now().After(deadline) {
			return JobResult{}, false
		}
		time.Sleep(2 * time.Millisecond)
	}
}
